src/detect-mqtt-qos.c

```diff
@@ -0,0 +1,289 @@
+        DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const Signature *s, const SigMatchData *smd,
+        Flow *f, uint8_t flags, void *alstate,
+        void *txv, uint64_t tx_id);
+
+/**
+ * \brief Registration function for mqtt.qos: keyword
+ */
+void DetectMQTTQosRegister (void)
+{
+    sigmatch_table[DETECT_AL_MQTT_QOS].name = "mqtt.qos";
+    sigmatch_table[DETECT_AL_MQTT_QOS].desc = "match MQTT fixed header QOS level";
+    sigmatch_table[DETECT_AL_MQTT_QOS].url = "/rules/mqtt-keywords.html#mqtt-qos";
+    sigmatch_table[DETECT_AL_MQTT_QOS].AppLayerTxMatch = DetectMQTTQosMatch;
+    sigmatch_table[DETECT_AL_MQTT_QOS].Setup = DetectMQTTQosSetup;
+    sigmatch_table[DETECT_AL_MQTT_QOS].Free  = DetectMQTTQosFree;
+    sigmatch_table[DETECT_AL_MQTT_QOS].RegisterTests = MQTTQosRegisterTests;
+
+    DetectSetupParseRegexes(PARSE_REGEX, &parse_regex);
+
+    DetectAppLayerInspectEngineRegister("mqtt.qos",
+            ALPROTO_MQTT, SIG_FLAG_TOSERVER, 1,
+            DetectEngineInspectMQTTQosGeneric);
+
+    mqtt_qos_id = DetectBufferTypeGetByName("mqtt.qos");
+}
+
+static int DetectEngineInspectMQTTQosGeneric(ThreadVars *tv,
+        DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const Signature *s, const SigMatchData *smd,
+        Flow *f, uint8_t flags, void *alstate,
+        void *txv, uint64_t tx_id)
+{
+    return DetectEngineInspectGenericList(tv, de_ctx, det_ctx, s, smd,
+                                          f, flags, alstate, txv, tx_id);
+}
+
+/**
+ * \internal
+ * \brief Function to match fixed header QOS field of an MQTT Tx
+ *
+ * \param det_ctx Pointer to the pattern matcher thread.
+ * \param f       Pointer to the current flow.
+ * \param flags   Flags.
+ * \param state   App layer state.
+ * \param txv     Pointer to the transaction.
+ * \param s       Pointer to the Signature.
+ * \param ctx     Pointer to the sigmatch that we will cast into uint8_t.
+ *
+ * \retval 0 no match.
+ * \retval 1 match.
+ */
+static int DetectMQTTQosMatch(DetectEngineThreadCtx *det_ctx,
+                               Flow *f, uint8_t flags, void *state,
+                               void *txv, const Signature *s,
+                               const SigMatchCtx *ctx)
+{
+    const uint8_t *de = (const uint8_t *)ctx;
+    uint8_t qosval;
+
+    if (!de)
+        return 0;
+
+    rs_mqtt_tx_get_qos(txv, &qosval);
+    if (qosval == *de)
+        return 1;
+
+    return 0;
+}
+
+/**
+ * \internal
+ * \brief This function is used to parse options passed via mqtt.qos: keyword
+ *
+ * \param rawstr Pointer to the user provided options
+ *
+ * \retval de pointer to DetectMQTTQosData on success
+ * \retval NULL on failure
+ */
+static uint8_t *DetectMQTTQosParse(const char *rawstr)
+{
+    uint8_t *de = NULL;
+#define MAX_SUBSTRINGS 30
+    int ret = 0;
+    uint8_t val;
+    int ov[MAX_SUBSTRINGS];
+
+    ret = DetectParsePcreExec(&parse_regex, rawstr, 0, 0, ov, MAX_SUBSTRINGS);
```

Comment:
PCRE seems an overkill here

Reply:
True. Removed that here and just worked with the parsed integer value.